pipelines/curate_coding.py

```
@@ -419,6 +478,496 @@ def curate_jsonl(source_path: str | Path) -> dict[str, Any]:
+            violations.append(
+                f"{step_where}: retained with impossible reason codes "
+                f"{sorted(impossible_reasons)}"
+            )
+        if action == "retained" and thought_fields_removed != 0:
+            violations.append(f"{step_where}: retained step reports thought removals")
+    else:
+        violations.append(f"{step_where}: unknown step action {action!r}")
+    return violations
+
+
+def verify_manifest(
```

Ruff (0.16.2)

[warning] Too many branches (43 > 12)

(PLR0912)